pkg/ring/lifecycler.go

```diff
@@ -588,6 +589,13 @@ func (i *Lifecycler) initRing(ctx context.Context) error {
 		i.setTokens(tokens)
 
 		level.Info(log.Logger).Log("msg", "existing entry found in ring", "state", i.GetState(), "tokens", len(tokens), "ring", i.RingName)
+
+		// If we flip the instance from LEAVING to ACTIVE and the heartbeat is disabled, we need to update KV here
+		if i.cfg.HeartbeatPeriod == 0 && ringDesc.Ingesters[i.ID].State == LEAVING && instanceDesc.State == ACTIVE {
+			ringDesc.Ingesters[i.ID] = instanceDesc
+			return ringDesc, true, nil
+		}
+
 		// we haven't modified the ring, don't try to store it.
 		return nil, true, nil
```

Comment:
I tried to address a similar (but not the same) issue in the PR https://github.com/cortexproject/cortex/pull/3774. Part of the changes there was to do the following here:

```
		// Update the ring if the instance has been changed.
		if !instanceDesc.Equal(ringDesc.Ingesters[i.ID]) {
			// Update timestamp to give gossiping client a chance register ring change.
			instanceDesc.Timestamp = time.Now().Unix()

			ringDesc.Ingesters[i.ID] = instanceDesc
			return ringDesc, true, nil
		}
```

My take was that it's not true that "we haven't modified the ring" once we reach this point. Actually, we have have modified it, but we don't update it in the ring.

Could you also take a look at #3774 and share your thoughts, please?

Reply:
I agree with that...
I just kept the comment/logic to no update the ring here to be conservative. The `heartbeat-period=0` is marked as experimental and the logic on this PR will only be trigged in this case - this way we can verify any potential side effect of it without impact on existing configs.

I will change my logic to be like this one you pointed out here but i wanna still keep the `i.cfg.HeartbeatPeriod == 0`, make sense?.

The difference i see btw here and #3774 is that there we were flipping from LEAVING to PENDING, right? On this PR we are just maintaining the same logic as before i guess.. just updating KV a little earlier.